tensorflow/compiler/mlir/tensorflow/transforms/region_control_flow_to_functional.cc

```diff
@@ -217,8 +217,18 @@ bool MatchCallArgs(CallOp first, CallOp second, ArgMatcherFn matcher) {
   for (auto it : llvm::zip(first.getArgOperands(), second.getArgOperands())) {
     // Get the defining Op, skipping over casts.
     auto get_defining_op = [](Value value) {
-      while (llvm::isa_and_nonnull<CastOp>(value.getDefiningOp()))
-        value = cast<CastOp>(value.getDefiningOp()).getOperand();
+      while (auto cast_op =
+                 llvm::dyn_cast_or_null<CastOp>(value.getDefiningOp())) {
+        // Consider cast compatibility in case
+        //    %cast = "tf.Cast"(%0) : (tensor<2xi64>) -> tensor<2xf32>
+        // is skipped.
+        if (AreCastCompatible(
```

Comment:
AreCastCompatible is used for shape verification. I think you want to verify that source element type of operand and result are the same.

Reply:
I think it checks both element type and shape and is used https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc#L792 to verfiy IfOp/WhileOp.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/tensorflow/ir/tf_types.cc#L251-L268

Reply:
We can expect all legal tensorflow tf.Cast ops to have the same operand and result shape at runtime. We don't have a verifier for this but we should add one.

Reply:
Got it!